### ISSUE TITLE
Limit DoSBlocks

### ISSUE THREAD

**Issue by lukechampine** [NebulousLabs/Sia#986]
The `dosBlocks` field of the consensus set is currently unbounded. This allows attackers to fill up target memory by spamming peers with DoS blocks.

**Comment by VoidingWarranties** [comment]
Interesting dilemma. If we limit the number of DoS blocks stored we need to decide which DoS blocks to kick out when a new one is found. LRU cache?

**Comment by lukechampine** [comment]
yes, that would probably be sufficient. Actually, it might be best to just replace at random. The goal of the attacker would be to spam you with bad blocks not in your `dosBlocks` set. If they can predict which blocks will be replaced, they can take advantage of that. I don't think random replacement is a perfect scheme either though. More forceful would be to ban any peer who sends you too many bad blocks.

**Comment by DavidVorick** [comment]
I would opt for banning peers. I'm not really sure how we would accomplish deleting at random, given that I'm planning on moving the dosBlocks field to a database. Even if it were in a map though, I'm still not sure how you would effectively select one at random to be deleted. You don't know what the keys are without iterating through the whole list, which is too expensive of an operation to be useful for stopping DoS attacks.
